pkg/services/ngalert/notifier/channels/slack.go

```diff
@@ -35,132 +35,110 @@ type SlackNotifier struct {
+	return ch, nil
 }
 
-func NewSlackConfig(factoryConfig FactoryConfig) (*SlackConfig, error) {
+func buildSlackNotifier(factoryConfig FactoryConfig) (*SlackNotifier, error) {
 	channelConfig := factoryConfig.Config
 	decryptFunc := factoryConfig.DecryptFunc
-	endpointURL := channelConfig.Settings.Get("endpointUrl").MustString(SlackAPIEndpoint)
-	slackURL := decryptFunc(context.Background(), channelConfig.SecureSettings, "url", channelConfig.Settings.Get("url").MustString())
+	var settings slackSettings
+	err := factoryConfig.Config.marshalSettings(&settings)
+	if err != nil {
+		return nil, err
+	}
+
+	if settings.EndpointURL == "" {
+		settings.EndpointURL = SlackAPIEndpoint
+	}
+	slackURL := decryptFunc(context.Background(), channelConfig.SecureSettings, "url", settings.URL)
 	if slackURL == "" {
-		slackURL = endpointURL
+		slackURL = settings.EndpointURL
 	}
 	apiURL, err := url.Parse(slackURL)
 	if err != nil {
 		return nil, fmt.Errorf("invalid URL %q", slackURL)
 	}
-	recipient := strings.TrimSpace(channelConfig.Settings.Get("recipient").MustString())
-	if recipient == "" && apiURL.String() == SlackAPIEndpoint {
+	settings.URL = apiURL.String()
+
+	settings.Recipient = strings.TrimSpace(settings.Recipient)
+	if settings.Recipient == "" && settings.URL == SlackAPIEndpoint {
 		return nil, errors.New("recipient must be specified when using the Slack chat API")
 	}
-	mentionChannel := channelConfig.Settings.Get("mentionChannel").MustString()
-	if mentionChannel != "" && mentionChannel != "here" && mentionChannel != "channel" {
-		return nil, fmt.Errorf("invalid value for mentionChannel: %q", mentionChannel)
+	if settings.MentionChannel != "" && settings.MentionChannel != "here" && settings.MentionChannel != "channel" {
+		return nil, fmt.Errorf("invalid value for mentionChannel: %q", settings.MentionChannel)
 	}
-	token := decryptFunc(context.Background(), channelConfig.SecureSettings, "token", channelConfig.Settings.Get("token").MustString())
-	if token == "" && apiURL.String() == SlackAPIEndpoint {
+	settings.Token = decryptFunc(context.Background(), channelConfig.SecureSettings, "token", settings.Token)
+	if settings.Token == "" && settings.URL == SlackAPIEndpoint {
 		return nil, errors.New("token must be specified when using the Slack chat API")
 	}
-	mentionUsersStr := channelConfig.Settings.Get("mentionUsers").MustString()
-	mentionUsers := []string{}
-	for _, u := range strings.Split(mentionUsersStr, ",") {
+	settings.MentionUsers = []string{}
+	for _, u := range strings.Split(settings.MentionUsersStr, ",") {
 		u = strings.TrimSpace(u)
 		if u != "" {
-			mentionUsers = append(mentionUsers, u)
+			settings.MentionUsers = append(settings.MentionUsers, u)
 		}
 	}
-	mentionGroupsStr := channelConfig.Settings.Get("mentionGroups").MustString()
-	mentionGroups := []string{}
-	for _, g := range strings.Split(mentionGroupsStr, ",") {
+	settings.MentionGroups = []string{}
+	for _, g := range strings.Split(settings.MentionGroupsStr, ",") {
```

Comment:
there is a `util.SplitString`. I think you can use it here.

Reply:
`util.SplitString` splits by spaces in addition to commas. I don't think we can use this for MentionUsers as Slack users can sometimes have spaces in them. We would stop accepting some slack 

MentionGroups should not have this problem. However, I'd like to be able to share the string parsing code as we did in your other example, but we can't do so if we move one to SplitString and the other not. I ended up sticking with the current comma-only split for this reason.